### PR TITLE
storelimit: impl fead back by snapshot report

### LINKED ISSUE #6147: add controller to adjust sliding windows capacity based on the snapshot statistics.

### PR DESCRIPTION
**Pull request by bufferflies**
### What problem does this PR solve?

Issue Number: Close #6147

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

Master

image

In this PR 
image

Code changes

Side effects

- Increased code complexity

Related changes

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/core/storelimit/limit_test.go [hunk 1]
```diff
@@ -108,3 +112,122 @@ func TestWindow(t *testing.T) {
...
+
+						// the tikv should contain two regions, one is running and another is waiting.
+						if iter < 5 {
+							re.Greater(s.GetCap(), regionSize)
+							re.Less(s.GetCap(), regionSize*wait)
+						}
+						s.Ack(regionSize, SendSnapshot)
+					}
+				}
+>>>>>>> 5689302dd9cc23dc18bdee9a05af220b4778b6b9
```
Comment (zhangjinpeng1987) [inline pkg/core/storelimit/limit_test.go:226]: please fix conflicts

Comment (bufferflies) [reply pkg/core/storelimit/limit_test.go]: fixed

File: pkg/core/storelimit/sliding_window.go [hunk 2]
```diff
@@ -62,13 +87,20 @@ func (s *SlidingWindows) Reset(cap float64, typ Type) {
...
+// GetCap returns the capacity of the sliding windows.
+func (s *SlidingWindows) GetCap() int64 {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.windows[0].capacity
+}
+
 // GetUsed returns the used size in the sliding windows.
-func (s *SlidingWindows) GetUsed() int64 {
+func (s *SlidingWindows) GetUsed() []int64 {
```
Comment (CabinfeverB) [inline pkg/core/storelimit/sliding_window.go:98]: `GetUsed ` is only used in metrics and test?

**Comment by CabinfeverB** [comment]
CI failed again.

File: pkg/core/storelimit/sliding_window.go (hunk 2 above)
Comment (bufferflies) [reply pkg/core/storelimit/sliding_window.go:98]: yes

File: pkg/core/store.go [hunk 3]
```diff
@@ -106,11 +106,27 @@ func (s *StoreInfo) Clone(opts ...StoreCreateOption) *StoreInfo {
 	store := *s
 	store.meta = typeutil.DeepClone(s.meta, StoreFactory)
 	for _, opt := range opts {
-		opt(&store)
+		if opt != nil {
```
Comment (rleungx) [inline pkg/core/store.go:109]: When will it become nil?

File: server/config/config.go [hunk 4]
```diff
@@ -777,6 +784,11 @@ func (c *ScheduleConfig) adjust(meta *configutil.ConfigMetaData, reloading bool)
 	if !meta.IsDefined("store-limit-mode") {
 		configutil.AdjustString(&c.StoreLimitMode, defaultStoreLimitMode)
 	}
+
+	if !meta.IsDefined("store-limit-version") {
```
Comment (rleungx) [inline server/config/config.go:788]: Do we need to consider reloading?

File: pkg/core/store.go [hunk 5]
```diff
@@ -106,11 +106,27 @@ func (s *StoreInfo) Clone(opts ...StoreCreateOption) *StoreInfo {
...
 
+// LimitVersion returns the limit version of the store.
+func (s *StoreInfo) LimitVersion() string {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.limiter.Version()
+}
+
+// Feedback is used to update the store's limit.
+func (s *StoreInfo) Feedback(e float64) {
```
Comment (rleungx) [inline pkg/core/store.go:124]: no need a lock?

File: pkg/statistics/store_collection.go [hunk 6]
```diff
@@ -111,6 +112,15 @@ func (s *storeStatistics) Observe(store *core.StoreInfo, stats *StoresStats) {
 	s.RegionCount += store.GetRegionCount()
 	s.LeaderCount += store.GetLeaderCount()
 	s.WitnessCount += store.GetWitnessCount()
+	limit, ok := store.GetStoreLimit().(*storelimit.SlidingWindows)
+	if ok {
+		cap := limit.GetCap()
+		storeStatusGauge.WithLabelValues(storeAddress, id, "windows_size").Set(float64(cap))
+		for i, use := range limit.GetUsed() {
+			storeStatusGauge.WithLabelValues(storeAddress, id, "windows_used_level_"+strconv.Itoa(i)).Set(float64(use))
```
Comment (rleungx) [inline pkg/statistics/store_collection.go:120]: Do we need to change the levels 0, 1, 2 to low, medium, high...?

File: server/config/config.go (hunk 4 above)
Comment (bufferflies) [reply server/config/config.go:788]: In this pr, `""` is default config. It should be considered if the default config is setted as `v2`

File: pkg/core/store.go (hunk 5 above)
Comment (bufferflies) [reply pkg/core/store.go:124]: It's thread safe in `limit.Feedback`

File: pkg/statistics/store_collection.go (hunk 6 above)
Comment (bufferflies) [reply pkg/statistics/store_collection.go:120]: good suggestion.

File: pkg/core/store.go (hunk 5 above)
Comment (rleungx) [reply pkg/core/store.go:124]: Does the limit itself can be modified?

File: server/cluster/cluster.go [hunk 7]
```diff
@@ -95,6 +95,10 @@ const (
 	removingAction          = "removing"
 	preparingAction         = "preparing"
 	gcTunerCheckCfgInterval = 10 * time.Second
+
+	// minTolerateDurationSec is the minimum duration that a store can tolerate.
```
Comment (rleungx) [inline server/cluster/cluster.go:99]: minTolerateDurationSec or minSnapshotDurationSec?

File: server/cluster/cluster.go [hunk 8]
```diff
@@ -848,17 +852,29 @@ func (c *RaftCluster) HandleStoreHeartbeat(heartbeat *pdpb.StoreHeartbeatRequest
 		return errors.Errorf("store %v not found", storeID)
 	}
 
+	limit := store.GetStoreLimit()
+	version := c.opt.GetStoreLimitVersion()
+	var op core.StoreCreateOption
```
Comment (rleungx) [inline server/cluster/cluster.go:857]: opt or option is better. op usually stands for operator

File: server/cluster/cluster.go [hunk 9]
```diff
@@ -919,7 +935,16 @@ func (c *RaftCluster) HandleStoreHeartbeat(heartbeat *pdpb.StoreHeartbeatRequest
...
 	}
-
+	for _, stat := range stats.GetSnapshotStats() {
+		// the duration of snapshot is the sum between the send and generate snapshot.
+		// notice: to enlarge the limit in time, we reset the executing duration when it less than the minSnapshotDurationSec.
+		dur := stat.GetSendDurationSec() + stat.GetGenerateDurationSec()
+		if dur < minSnapshotDurationSec {
+			dur = minSnapshotDurationSec
+		}
+		e := int64(dur)*2 - int64(stat.GetTotalDurationSec())
```
Comment (rleungx) [inline server/cluster/cluster.go:945]: Please add some comments to explain how do we calculate the e?

File: pkg/core/storelimit/sliding_window.go [hunk 10]
```diff
@@ -30,30 +35,50 @@ var _ StoreLimit = &SlidingWindows{}
...
+// Feedback is used to update the capacity of the sliding windows.
+func (s *SlidingWindows) Feedback(e float64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	// If the limiter is available, we don't need to update the capacity.
+	if s.windows[constant.Low].available() {
+		return
+	}
+	s.lastSum += e
+	cap := defaultProportion*e + defaultIntegral*s.lastSum
```
Comment (rleungx) [inline pkg/core/storelimit/sliding_window.go:66]: ditto

File: pkg/core/store.go (hunk 5 above)
Comment (bufferflies) [reply pkg/core/store.go:124]: yes,

File: pkg/core/store.go (hunk 3 above)
Comment (bufferflies) [reply pkg/core/store.go:109]: https://github.com/tikv/pd/blob/ef2a01f7bcacd4074541cf4131c8a8f667a0604e/server/cluster/cluster.go#L872 in this case, the option maybe nil

Comment (bufferflies) [reply pkg/core/store.go:109]: https://github.com/tikv/pd/blob/ef2a01f7bcacd4074541cf4131c8a8f667a0604e/server/cluster/cluster.go#L872 in this case, the option maybe nil

File: server/cluster/cluster.go (hunk 9 above)
Comment (bufferflies) [reply server/cluster/cluster.go:953]: fixed

File: server/cluster/cluster.go (hunk 7 above)
Comment (bufferflies) [reply server/cluster/cluster.go]: fixed

File: server/cluster/cluster.go (hunk 8 above)
Comment (bufferflies) [reply server/cluster/cluster.go]: fixed

File: pkg/core/storelimit/sliding_window.go (hunk 10 above)
Comment (bufferflies) [reply pkg/core/storelimit/sliding_window.go:69]: fixed

File: server/cluster/cluster.go [hunk 11]
```diff
@@ -919,7 +935,24 @@ func (c *RaftCluster) HandleStoreHeartbeat(heartbeat *pdpb.StoreHeartbeatRequest
...
+		if dur < minSnapshotDurationSec {
+			dur = minSnapshotDurationSec
+		}
+		// This error is the diff between the executing duration and the waiting duration.
+		// The waiting duration is the total duration minus the executing duration.
+		// so e=executing_duration-waiting_duration=executing_duration-(total_duration-executing_duration)=2*executing_duration-total_duration
+		// Eg: the total duration is 20s, the executing duration is 10s, the error is 0s.
+		// Eg: the total duration is 20s, the executing duration is 8s, the error is -4s.
+		// Eg: the total duration is 10s, the executing duration is 12s, the error is 4s.
+		// if error is positive, it means the mot time cost in executing, pd should send more snapshot to this tikv.
```
Comment (HuSharp) [inline server/cluster/cluster.go:951]:
```suggestion
		// if error is positive, it means the most time cost in executing, pd should send more snapshot to this tikv.
```

File: pkg/core/storelimit/sliding_window.go [hunk 12]
```diff
@@ -30,30 +35,53 @@ var _ StoreLimit = &SlidingWindows{}
...
+		cap = defaultWindowSize
+	}
+	s.set(cap, SendSnapshot)
+}
+
+// Reset does nothing because the capacity depends on the feedback.
+func (s *SlidingWindows) Reset(_ float64, _ Type) {
+}
+
+func (s *SlidingWindows) set(cap float64, typ Type) {
```
Comment (HuSharp) [inline pkg/core/storelimit/sliding_window.go:81]: How about called it `resetAllCaps`?

File: server/cluster/cluster.go [hunk 13]
```diff
@@ -848,17 +852,29 @@ func (c *RaftCluster) HandleStoreHeartbeat(heartbeat *pdpb.StoreHeartbeatRequest
 		return errors.Errorf("store %v not found", storeID)
 	}
 
+	limit := store.GetStoreLimit()
+	version := c.opt.GetStoreLimitVersion()
+	var opt core.StoreCreateOption
+	if limit == nil || limit.Version() != version {
+		if version == storelimit.VersionV2 {
+			limit = storelimit.NewSlidingWindows()
```
Comment (HuSharp) [inline server/cluster/cluster.go:860]: How about just call `opt := core.SetStoreLimit(limit)`

File: server/cluster/cluster.go (hunk 11 above)
Comment (rleungx) [inline server/cluster/cluster.go:951]:
```suggestion
		// if error is positive, it means the most time cost in executing, pd should send more snapshot to this tikv.
```

**Review by rleungx — APPROVED** [review]

**Review by HuSharp — COMMENTED** [review]
BTW, does `TestFeedback` is a flaky test?

File: pkg/core/storelimit/sliding_window.go [hunk 14]
```diff
@@ -30,30 +35,53 @@ var _ StoreLimit = &SlidingWindows{}
...
+	defer s.mu.Unlock()
+	// If the limiter is available, we don't need to update the capacity.
+	if s.windows[constant.Low].available() {
+		return
+	}
+	s.lastSum += e
+	// There are two constants to control the proportion of the sum and the current error.
+	// The sum of the error is used to ensure the capacity is more stable even if the error is zero.
+	// In the final scene, the sum of the error should be stable and the current error should be zero.
+	cap := defaultProportion*e + defaultIntegral*s.lastSum
```
Comment (CabinfeverB) [inline pkg/core/storelimit/sliding_window.go:69]: If executing_duration is always greater than waiting_duration, will the cap will go up infinitely?

**Review by CabinfeverB — APPROVED** [review]
rest LGTM

**Comment by bufferflies** [comment]
> 

yes , I will fix it by using logical time.

File: pkg/core/storelimit/sliding_window.go (hunk 14 above)
Comment (bufferflies) [reply pkg/core/storelimit/sliding_window.go:69]: no, the cap will only go up iff the space is not available.

File: server/cluster/cluster.go (hunk 11 above)
Comment (bufferflies) [reply server/cluster/cluster.go]: fixed

File: pkg/core/storelimit/sliding_window.go (hunk 12 above)
Comment (bufferflies) [reply pkg/core/storelimit/sliding_window.go:81]: no, the caps should be set togther and not allowed to set them alone.

File: pkg/core/storelimit/sliding_window.go [hunk 15]
```diff
@@ -30,30 +35,50 @@ var _ StoreLimit = &SlidingWindows{}
...
+// Version returns v2
+func (s *SlidingWindows) Version() string {
+	return VersionV2
+}
+
+// Feedback is used to update the capacity of the sliding windows.
+func (s *SlidingWindows) Feedback(e float64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	// If the limiter is not available, we don't need to update the capacity.
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply pkg/core/storelimit/sliding_window.go]: fixed, the comment should be ` If the limiter is available, we don't need to update the capacity.`

**Pull request merged**